File: indexrecyclerview/src/main/java/com/banzhi/indexrecyclerview/IndexBar.java
package com.banzhi.indexrecyclerview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;

import com.banzhi.indexrecyclerview.bean.BaseIndexBean;
import com.banzhi.indexrecyclerview.utils.IndexDataHelper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * <pre>
 * @author : No.1
 * @time : 2018/7/26.
 * @desciption : 索引
 * @version :
 * </pre>
 */

public class IndexBar extends View {
    private static final String TAG = "tag";

    public static final int VERTICAL = 1;
    public static final int HORIZONTAL = 2;

    private Context mContext;
    /**
     * 默认索引
     */
    private static final String[] DEFAULT_INDEX = new String[]{"A", "B", "C", "D", "E", "F", "G", "H", "I",
            "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V",
            "W", "X", "Y", "Z", "#"};

    /**
     * 每个index的高度
     */
    private int indexHeght;
    /**
     * view宽度
     */
    private int mWidth;
    /**
     * view高度
     */
    private int mHeight;
    /**
     * 画笔
     */
    private Paint mPaint;
    /**
     * 按下时的背景颜色
     */
    private int mPressColor;
    /**
     * 按下时文字的颜色
     */
    private int mTextColor;
    /**
     * 文字的颜色
     */
    private int mPressTextColor;
    /**
     * 字体大小
     */
    private int textSize;

    private int DEFAULT_PRESS_COLOR = Color.GRAY;

    List<String> indexDatas;
    /**
     * 使用数据内容作为索引
     */
    boolean useDatasIndex;
    /**
     * 数据是否有序
     */
    boolean isOrderly;

    /**
     * 临时保存view背景颜色
     */
    private int color = android.R.color.transparent;
    /**
     * indexbar方向
     */
    private int mOrientation;

    RecyclerView.LayoutManager layoutManager;

    public IndexBar(Context context) {
        super(context);
        init(context, null, -1);

    }

    public IndexBar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, -1);
    }

    public IndexBar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        this.mContext = context;
        //默认的TextSize
        int DEFAULT_SIZE = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 16, getResources().getDisplayMetrics());
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.IndexrecyclerviewIndexBar);
        if (typedArray != null) {
            textSize = typedArray.getDimensionPixelSize(R.styleable.IndexrecyclerviewIndexBar_textSize, DEFAULT_SIZE);
            mPressColor = typedArray.getColor(R.styleable.IndexrecyclerviewIndexBar_pressColor, DEFAULT_PRESS_COLOR);
            mPressTextColor = typedArray.getColor(R.styleable.IndexrecyclerviewIndexBar_pressTextColor, DEFAULT_PRESS_COLOR);
            mTextColor = typedArray.getColor(R.styleable.IndexrecyclerviewIndexBar_textColor, DEFAULT_PRESS_COLOR);
            mOrientation = typedArray.getInt(R.styleable.IndexrecyclerviewIndexBar_orientation, VERTICAL);
        }
        initPaint();
        initDatas();
        setOnIndexPressListener(new OnIndexPressListener() {
            @Override
            public void onIndexChange(int index, String text) {
                textView.setText(text);
                textView.setVisibility(VISIBLE);
                if (layoutManager != null) {
                    int position = getPosByTag(text);
                    if (position > -1) {
                        ((LinearLayoutManager) layoutManager).scrollToPositionWithOffset(position, 0);
                    }
                }
            }

            @Override
            public void onMotionEventEnd() {
                textView.setVisibility(GONE);
            }
        });
    }

    TextView textView;

    public void setTextView(TextView textView) {
        this.textView = textView;
    }

    RecyclerView mRecyclerView;
    int mHeadCount;

    public void bindRecyclerView(RecyclerView recyclerView) {
        this.mRecyclerView = recyclerView;
        if (mRecyclerView != null) {
            layoutManager = mRecyclerView.getLayoutManager();
            RecyclerView.ItemDecoration itemDecoration = mRecyclerView.getItemDecorationAt(0);
            if (itemDecoration instanceof LevitationDecoration) {
                mHeadCount = ((LevitationDecoration) itemDecoration).getHeadCount();
            }
        }
    }

    private void initDatas() {
        if (!useDatasIndex) {
            indexDatas = Arrays.asList(DEFAULT_INDEX);
        } else {
            indexDatas = new ArrayList<>();
        }
    }

    private void initPaint() {
        mPaint = new Paint();
        mPaint.setTextSize(textSize);
        mPaint.setAntiAlias(true);
    }

    @SuppressLint("ResourceAsColor")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            //手指抬起时背景恢复透明
            Drawable background = getBackground();
            if (background != null) {
                color = ((ColorDrawable) background).getColor();
            }
            setBackgroundColor(mPressColor);
            computePressIndexLocation(event.getX(), event.getY());

        } else if (event.getAction() == MotionEvent.ACTION_MOVE) {
            computePressIndexLocation(event.getX(), event.getY());
        } else {
            //手指抬起时背景恢复透明
            setBackgroundColor(color);
            //重置当前位置
            currentIndex = -1;

            if (mOnIndexPressListener != null) {
                mOnIndexPressListener.onMotionEventEnd();
            }
        }
        return true;
    }

    private int currentIndex = -1;

    /**
     * 计算按下的位置
     */
    private void computePressIndexLocation(float x, float y) {
        if (mOrientation == VERTICAL) {
            // 计算按下的区域位置
            currentIndex = (int) ((y - getPaddingTop()) / indexHeght);
        } else {
            currentIndex = (int) ((x - getPaddingLeft()) / indexHeght);
        }
        if (currentIndex < 0) {
            currentIndex = 0;
        } else if (currentIndex >= indexDatas.size()) {
            currentIndex = indexDatas.size() - 1;
        }
        invalidateMySelft();
        if (mOnIndexPressListener != null) {
            mOnIndexPressListener.onIndexChange(currentIndex, indexDatas.get(currentIndex));
        }

    }

    /**
     * 源数据时候有序
     *
     * @param orderly
     */
    public void setOrderly(boolean orderly) {
        isOrderly = orderly;
    }

    /**
     * 使用源数据作为索引
     */
    public void setUseDatasIndex() {
        this.useDatasIndex = true;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mWidth = w;
        mHeight = h;
        computeIndexHeight();
    }

    /**
     * 计算单个index高度
     */
    private void computeIndexHeight() {
        if (mOrientation == VERTICAL) {
            indexHeght = (mHeight - getPaddingTop() - getPaddingBottom()) / indexDatas.size();
        } else {
            indexHeght = (mWidth - getPaddingLeft() - getPaddingRight()) / indexDatas.size();
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        for (int i = 0; i < indexDatas.size(); i++) {
            String index = indexDatas.get(i);
            Paint.FontMetrics metrics = mPaint.getFontMetrics();
            //计算baseline
            int baseLine = (int) ((indexHeght - metrics.bottom - metrics.top) / 2);
            if (currentIndex == i) {
                mPaint.setColor(mPressTextColor);
            } else {
                mPaint.setColor(mTextColor);
            }
            //绘制文字
            if (mOrientation == VERTICAL) {
                canvas.drawText(index, mWidth / 2 - mPaint.measureText(index) / 2,
                        getPaddingTop() + baseLine + indexHeght * i, mPaint);
            } else {
                canvas.drawText(index, getPaddingLeft() + indexHeght * i + mPaint.measureText(index) / 2, mHeight - getPaddingBottom()
                        , mPaint);
            }
            Log.i(TAG, "onDraw: x=" + (getPaddingLeft() + indexHeght * i) + "  y=" + (mHeight));
        }

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        //取出宽高的MeasureSpec  Mode 和Size
        int wMode = MeasureSpec.getMode(widthMeasureSpec);
        int wSize = MeasureSpec.getSize(widthMeasureSpec);
        int hMode = MeasureSpec.getMode(heightMeasureSpec);
        int hSize = MeasureSpec.getSize(heightMeasureSpec);
        //最终测量出来的宽高
        int measureWidth = 0, measureHeight = 0;

        //得到合适宽度：
        //存放每个绘制的index的Rect区域
        Rect indexBounds = new Rect();
        String index;//每个要绘制的index内容
        for (int i = 0; i < indexDatas.size(); i++) {
            index = indexDatas.get(i);
            //测量计算文字所在矩形，可以得到宽高
            mPaint.getTextBounds(index, 0, index.length(), indexBounds);
            //循环结束后，得到index的最大宽度
            if (mOrientation == VERTICAL) {
                measureWidth = Math.max(indexBounds.width() + getPaddingLeft() + getPaddingRight(), measureWidth);
            } else {
                measureWidth = Math.max(indexBounds.width(), measureWidth);
            }
            //循环结束后，得到index的最大高度，然后*size
            if (mOrientation == VERTICAL) {
                measureHeight = Math.max(indexBounds.height(), measureHeight);
            } else {
                measureHeight = Math.max(indexBounds.height() + getPaddingTop() + getPaddingBottom(), measureHeight);
            }
            Log.i(TAG, "onMeasure: index=" + index + "  width=" + indexBounds.width() + "  height=" + indexBounds.height()
                    + "  measureWidth=" + measureWidth + "  measureHeight=" + measureHeight);

        }
        if (mOrientation == VERTICAL) {
            measureHeight *= indexDatas.size();
        } else {
            measureWidth *= indexDatas.size();
        }
        if (wMode == MeasureSpec.EXACTLY) {
            measureWidth = wSize;
        } else if (wMode == MeasureSpec.AT_MOST) {
            //wSize此时是父控件能给子View分配的最大空间
            measureWidth = Math.min(measureWidth, wSize);
        } else if (wMode == MeasureSpec.UNSPECIFIED) {

        }
        //得到合适的高度：
        if (hMode == MeasureSpec.EXACTLY) {
            measureHeight = hSize;
        } else if (hMode == MeasureSpec.AT_MOST) {
            //wSize此时是父控件能给子View分配的最大空间
            measureHeight = Math.min(measureHeight, hSize);
        } else if (hMode == MeasureSpec.UNSPECIFIED) {

        }
        setMeasuredDimension(measureWidth, measureHeight);
    }

    OnIndexPressListener mOnIndexPressListener;

    public void setOnIndexPressListener(OnIndexPressListener mOnIndexPressListener) {
        this.mOnIndexPressListener = mOnIndexPressListener;
    }

    public interface OnIndexPressListener {
        /**
         * @param index 当前选中的位置
         * @param text  选中的文字
         */
        void onIndexChange(int index, String text);

        /**
         * 事件结束时回调
         */
        void onMotionEventEnd();
    }

    /**
     * 原始数据
     */
    List<? extends BaseIndexBean> sourceDatas;

    /**
     * 设置原始数据
     *
     * @param sourceDatas
     */
    public void setSourceDatas(List<? extends BaseIndexBean> sourceDatas) {
        this.sourceDatas = sourceDatas;
        initIndexDatas();
        invalidateMySelft();

    }

    private void invalidateMySelft() {
        if (isMainThread()) {
            invalidate();
        } else {
            postInvalidate();
        }
    }

    public boolean isMainThread() {
        return Thread.currentThread() == Looper.getMainLooper().getThread();
    }

    /**
     * 初始原始数据 并提取索引
     */
    private void initIndexDatas() {
        if (null == sourceDatas || sourceDatas.isEmpty()) {
            return;
        }
        IndexDataHelper dataHelper = new IndexDataHelper();
        dataHelper.cover(sourceDatas);
        //源数据无序
        if (!isOrderly) {
            dataHelper.sortDatas(sourceDatas);
        }
        if (useDatasIndex) {
            indexDatas = new ArrayList<>();
            dataHelper.getIndex(sourceDatas, indexDatas);
            computeIndexHeight();
        }
    }

    /**
     * 根据传入的tag返回position
     *
     * @param tag
     * @return
     */
    private int getPosByTag(String tag) {
        if (null == sourceDatas || sourceDatas.isEmpty()) {
            return -1;
        }
        if (TextUtils.isEmpty(tag)) {
            return -1;
        }
        for (int i = 0; i < sourceDatas.size(); i++) {
            if (tag.equals(sourceDatas.get(i).getIndexTag())) {
                return i + mHeadCount;
            }
        }
        return -1;
    }
}

File: indexrecyclerview/src/main/java/com/banzhi/indexrecyclerview/widget/IndexBar.kt
package com.banzhi.indexrecyclerview.widget

import android.R
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.os.Looper
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.banzhi.indexrecyclerview.bean.BaseIndexBean
import com.banzhi.indexrecyclerview.decoration.LevitationDecoration
import com.banzhi.indexrecyclerview.interfaces.IDataHelper
import com.banzhi.indexrecyclerview.utils.IndexDataHelper
import java.util.*

/**
 * <pre>
 * @author :
 * @time : 2021/5/31.
 * @desciption : 索引
 * @version :
</pre> *
 */
class IndexBar : View {
    private var mContext: Context? = null

    /**
     * 每个index的高度
     */
    private var indexHeght = 0

    /**
     * view宽度
     */
    private var mWidth = 0

    /**
     * view高度
     */
    private var mHeight = 0

    /**
     * 画笔
     */
    private var mPaint: Paint? = null

    /**
     * 按下时的背景颜色
     */
    private var mPressBackground = 0

    /**
     * 文字颜色
     */
    private var mTextColor = 0

    /**
     * 按下时文字的颜色
     */
    private var mPressTextColor = 0

    /**
     * 文字选中的颜色
     */
    private var mSelectTextColor = 0

    /**
     * 字体大小
     */
    private var textSize = 0
    private val DEFAULT_PRESS_COLOR = Color.GRAY
    private val DEFAULT_BACKGROUND = Color.TRANSPARENT
    private var indexDatas: MutableList<String>? = null

    /**
     * 使用数据内容作为索引
     */
    private var useDatasIndex = false

    /**
     * 数据是否有序
     */
    private var isOrderly = false

    /**
     * 临时保存view背景颜色
     */
    private var color = R.color.transparent

    /**
     * indexbar方向
     */
    private var mOrientation = 0

    /**
     * 数据转换帮助类
     */
    private var mDataHelper: IDataHelper? = null
    private var layoutManager: RecyclerView.LayoutManager? = null

    constructor(context: Context) : super(context) {
        init(context, null, -1)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs, -1)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context, attrs, defStyleAttr)
    }

    private fun init(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
        mContext = context
        //默认的TextSize
        val DEFAULT_SIZE =
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, 16f, resources.displayMetrics)
                .toInt()
        val typedArray = context.obtainStyledAttributes(
            attrs,
            com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar
        )
        if (typedArray != null) {
            textSize = typedArray.getDimensionPixelSize(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_indexTextSize,
                DEFAULT_SIZE
            )
            mPressBackground = typedArray.getColor(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_pressBackground,
                DEFAULT_BACKGROUND
            )
            mTextColor = typedArray.getColor(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_indexTextColor,
                DEFAULT_PRESS_COLOR
            )
            mPressTextColor = typedArray.getColor(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_pressTextColor,
                mTextColor
            )
            mSelectTextColor = typedArray.getColor(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_selectTextColor,
                mTextColor
            )
            mOrientation = typedArray.getInt(
                com.banzhi.indexrecyclerview.R.styleable.IndexrecyclerviewIndexBar_orientation,
                VERTICAL
            )
        }
        initPaint()
        initDatas()
        setOnIndexPressListener(object : OnIndexPressListener {
            override fun onIndexChange(index: Int, text: String) {
                if (textView != null) {
                    textView!!.text = text
                    textView!!.visibility = VISIBLE
                }
                if (layoutManager != null) {
                    val position = getPosByTag(text)
                    Log.i(TAG, "onIndexChange: position===>$position")
                    if (position > -1) {
                        (layoutManager as LinearLayoutManager).scrollToPositionWithOffset(
                            position,
                            0
                        )
                    }
                }
            }

            override fun onMotionEventEnd() {
                if (textView != null) {
                    textView!!.visibility = GONE
                }
            }
        })
    }

    private var textView: TextView? = null
    fun setTextView(textView: TextView?) {
        this.textView = textView
    }

    var mRecyclerView: RecyclerView? = null
    var mHeadCount = 0
    fun bindRecyclerView(recyclerView: RecyclerView?) {
        mRecyclerView = recyclerView
        if (mRecyclerView != null) {
            layoutManager = mRecyclerView!!.layoutManager
            val itemDecoration = mRecyclerView!!.getItemDecorationAt(0)
            if (itemDecoration is LevitationDecoration) {
                mHeadCount = itemDecoration.headCount
            }
        }
    }

    private fun initDatas() {
        indexDatas = if (!useDatasIndex) {
            Arrays.asList(*DEFAULT_INDEX)
        } else {
            ArrayList()
        }
    }

    private fun initPaint() {
        mPaint = Paint()
        mPaint!!.textSize = textSize.toFloat()
        mPaint!!.isAntiAlias = true
    }

    var isPress = false

    @SuppressLint("ResourceAsColor")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_DOWN) {
            isPress = true
            val background = background
            if (background != null) {
                color = (background as ColorDrawable).color
            }
            setBackgroundColor(mPressBackground)
            computePressIndexLocation(event.x, event.y)
        } else if (event.action == MotionEvent.ACTION_MOVE) {
            computePressIndexLocation(event.x, event.y)
        } else {
            isPress = false
            //手指抬起时背景恢复透明
            setBackgroundColor(color)
            //重置当前位置
            currentIndex = -1
            if (mOnIndexPressListener != null) {
                mOnIndexPressListener!!.onMotionEventEnd()
            }
        }
        return true
    }

    private var currentIndex = -1

    /**
     * 计算按下的位置
     */
    private fun computePressIndexLocation(x: Float, y: Float) {
        currentIndex = if (mOrientation == VERTICAL) {
            // 计算按下的区域位置
            ((y - paddingTop) / indexHeght).toInt()
        } else {
            ((x - paddingLeft) / indexHeght).toInt()
        }
        if (currentIndex < 0) {
            currentIndex = 0
        } else if (currentIndex >= indexDatas!!.size) {
            currentIndex = indexDatas!!.size - 1
        }
        invalidateMySelft()
        if (mOnIndexPressListener != null) {
            mOnIndexPressListener!!.onIndexChange(currentIndex, indexDatas!![currentIndex])
        }
    }

    /**
     * 源数据时候有序
     *
     * @param orderly
     */
    fun setOrderly(orderly: Boolean) {
        isOrderly = orderly
    }

    /**
     * 使用源数据作为索引
     */
    fun setUseDatasIndex() {
        useDatasIndex = true
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        mWidth = w
        mHeight = h
        computeIndexHeight()
    }

    /**
     * 计算单个index高度
     */
    private fun computeIndexHeight() {
        indexHeght = if (mOrientation == VERTICAL) {
            (mHeight - paddingTop - paddingBottom) / indexDatas!!.size
        } else {
            (mWidth - paddingLeft - paddingRight) / indexDatas!!.size
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        for (i in indexDatas!!.indices) {
            val index = indexDatas!![i]
            val metrics = mPaint!!.fontMetrics
            //计算baseline
            val baseLine = ((indexHeght - metrics.bottom - metrics.top) / 2).toInt()
            if (currentIndex == i) {
                mPaint!!.color = mSelectTextColor
            } else {
                mPaint!!.color = if (isPress) mPressTextColor else mTextColor
            }
            //绘制文字
            if (mOrientation == VERTICAL) {
                canvas.drawText(
                    index, mWidth / 2 - mPaint!!.measureText(index) / 2, (
                            paddingTop + baseLine + indexHeght * i).toFloat(), mPaint
                )
            } else {
                canvas.drawText(
                    index,
                    paddingLeft + indexHeght * i + mPaint!!.measureText(index) / 2,
                    (mHeight - paddingBottom
                            ).toFloat(),
                    mPaint
                )
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        //取出宽高的MeasureSpec  Mode 和Size
        val wMode = MeasureSpec.getMode(widthMeasureSpec)
        val wSize = MeasureSpec.getSize(widthMeasureSpec)
        val hMode = MeasureSpec.getMode(heightMeasureSpec)
        val hSize = MeasureSpec.getSize(heightMeasureSpec)
        //最终测量出来的宽高
        var measureWidth = 0
        var measureHeight = 0

        //得到合适宽度：
        //存放每个绘制的index的Rect区域
        val indexBounds = Rect()
        var index: String //每个要绘制的index内容
        for (i in indexDatas!!.indices) {
            index = indexDatas!![i]
            //测量计算文字所在矩形，可以得到宽高
            mPaint!!.getTextBounds(index, 0, index.length, indexBounds)
            //循环结束后，得到index的最大宽度
            measureWidth = if (mOrientation == VERTICAL) {
                Math.max(indexBounds.width() + paddingLeft + paddingRight, measureWidth)
            } else {
                Math.max(indexBounds.width(), measureWidth)
            }
            //循环结束后，得到index的最大高度，然后*size
            measureHeight = if (mOrientation == VERTICAL) {
                Math.max(indexBounds.height(), measureHeight)
            } else {
                Math.max(indexBounds.height() + paddingTop + paddingBottom, measureHeight)
            }
        }
        if (mOrientation == VERTICAL) {
            measureHeight *= indexDatas!!.size
        } else {
            measureWidth *= indexDatas!!.size
        }
        if (wMode == MeasureSpec.EXACTLY) {
            measureWidth = wSize
        } else if (wMode == MeasureSpec.AT_MOST) {
            //wSize此时是父控件能给子View分配的最大空间
            measureWidth = Math.min(measureWidth, wSize)
        } else if (wMode == MeasureSpec.UNSPECIFIED) {
        }
        //得到合适的高度：
        if (hMode == MeasureSpec.EXACTLY) {
            measureHeight = hSize
        } else if (hMode == MeasureSpec.AT_MOST) {
            //wSize此时是父控件能给子View分配的最大空间
            measureHeight = Math.min(measureHeight, hSize)
        } else if (hMode == MeasureSpec.UNSPECIFIED) {
        }
        setMeasuredDimension(measureWidth, measureHeight)
    }

    var mOnIndexPressListener: OnIndexPressListener? = null
    fun setOnIndexPressListener(mOnIndexPressListener: OnIndexPressListener?) {
        this.mOnIndexPressListener = mOnIndexPressListener
    }

    interface OnIndexPressListener {
        /**
         * @param index 当前选中的位置
         * @param text  选中的文字
         */
        fun onIndexChange(index: Int, text: String)

        /**
         * 事件结束时回调
         */
        fun onMotionEventEnd()
    }

    /**
     * 原始数据
     */
    private var sourceDatas: MutableList<out BaseIndexBean>? = null

    /**
     * 设置原始数据
     *
     * @param sourceDatas
     */
    fun setSourceDatas(sourceDatas: MutableList<out BaseIndexBean>?) {
        this.sourceDatas = sourceDatas
        initIndexDatas()
        invalidateMySelft()
    }

    private fun invalidateMySelft() {
        if (isMainThread) {
            invalidate()
        } else {
            postInvalidate()
        }
    }

    val isMainThread: Boolean
        get() = Thread.currentThread() === Looper.getMainLooper().thread

    /**
     * 初始原始数据 并提取索引
     */
    private fun initIndexDatas() {
        if (sourceDatas.isNullOrEmpty()) {
            return
        }
        if (mDataHelper == null) {
            mDataHelper = IndexDataHelper()
        }
        mDataHelper!!.cover(sourceDatas!!)
        //源数据无序
        if (!isOrderly) {
            mDataHelper!!.sortDatas(sourceDatas!!)
        }
        if (useDatasIndex) {
            indexDatas = mutableListOf()
            mDataHelper!!.getIndex(sourceDatas!!, indexDatas!!)
            computeIndexHeight()
        }
    }

    /**
     * 根据传入的tag返回position
     *
     * @param tag
     * @return
     */
    private fun getPosByTag(tag: String): Int {
        if (null == sourceDatas || sourceDatas!!.isEmpty()) {
            return -1
        }
        if (TextUtils.isEmpty(tag)) {
            return -1
        }
        for (i in sourceDatas!!.indices) {
            if (tag == sourceDatas!![i]!!.getIndexTag()) {
                return i + mHeadCount
            }
        }
        return -1
    }

    companion object {
        private const val TAG = "tag"
        const val VERTICAL = 1
        const val HORIZONTAL = 2

        /**
         * 默认索引
         */
        private val DEFAULT_INDEX = arrayOf(
            "A", "B", "C", "D", "E", "F", "G", "H", "I",
            "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V",
            "W", "X", "Y", "Z", "#"
        )
    }
}